python/fusion_engine_client/parsers/mixed_log_reader.py

```diff
@@ -0,0 +1,342 @@
+
+                self.valid_count += 1
+                self.message_counts.setdefault(header.message_type, 0)
+                self.message_counts[header.message_type] += 1
+
+                # Check for sequence number gaps.
+                if self.prev_sequence_number is not None and \
+                   (header.sequence_number - self.prev_sequence_number) != 1 and \
+                   not (header.sequence_number == 0 and self.prev_sequence_number == 0xFFFFFFFF):
+                    func = self.logger.warning if self.warn_on_gaps else self.logger.debug
+                    func('Data gap detected @ %d (0x%x). [sequence=%d, gap_size=%d, total_messages=%d]' %
+                         (offset_bytes, offset_bytes, header.sequence_number,
+                          header.sequence_number - self.prev_sequence_number, self.valid_count + 1))
+                self.prev_sequence_number = header.sequence_number
+
+                # Deserialize the payload if we need it.
+                need_payload = self.return_payload or self.index_builder is not None or self.time_range is not None
+                if need_payload:
+                    cls = message_type_to_class.get(header.message_type, None)
+                    if cls is not None:
+                        try:
+                            payload = cls()
+                            payload.unpack(buffer=payload_bytes, offset=0)
+                        except Exception as e:
+                            self.logger.error("Error parsing %s message: %s" % (header.get_type_string(), str(e)))
+                            payload = None
+                    else:
+                        payload = None
+
+                # Add this message to the index file.
+                if self.index_builder is not None:
+                    p1_time = payload.get_p1_time() if payload is not None else None
+                    self.index_builder.append(message_type=header.message_type, offset_bytes=offset_bytes,
+                                              p1_time=p1_time)
+
+                # Now, if this message is not in the user-specified filter criteria, skip it.
+                if self.message_types is not None and header.message_type not in self.message_types:
+                    self.logger.trace("Message type not requested. Skipping.", depth=1)
+                    continue
+                elif self.time_range is not None and not self.time_range.is_in_range(payload):
+                    if self.time_range.in_range_started() and self.index_builder is None:
+                        self.logger.debug("End of time range reached. Finished processing.")
+                        break
+                    else:
+                        self.logger.trace("Message not in time range. Skipping.", depth=1)
+                        continue
+
+                # Construct the result. If we're returning the payload, deserialize the payload.
+                result = []
+                if self.return_header:
```

Comment:
Possibly too late to change this, but I find it a lot easier to consume output from these mutli-optional output return types if they return a dict instead of an array or tuple. The consumers don't need to check the full set of enabled outputs, and don't need to be changed if you change the return values.

Reply:
Yeah, I tend to agree generally with the caveats that:
1. Technically returning a dict and keying into it by strings is less efficient than indexing into a tuple
2. Returning a tuple is more consistent with a lot of `for a, b, c in generator` types of patterns

Probably not too easy to change this one at this point, but for other similar stuff we can definitely consider using a dict instead.